spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -603,101 +636,239 @@ private void invokeBatchListener(final ConsumerRecords<K, V> records) {
+						this.acks.put(record);
 					}
-					try {
-						this.batchErrorHandler.handle(e, records, this.consumer);
+					if (producer != null) {
+						sendOffsetsToTransaction(producer);
 					}
-					catch (Exception ee) {
-						this.logger.error("Error handler threw an exception", ee);
+				}
+			}
+			catch (RuntimeException e) {
+				if (this.containerProperties.isAckOnError() && !this.autoCommit) {
+					for (ConsumerRecord<K, V> record : getHighestOffsetRecords(recordList)) {
+						this.acks.add(record);
 					}
-					catch (Error er) { //NOSONAR
-						this.logger.error("Error handler threw an error", er);
-						throw er;
+				}
+				if (this.batchErrorHandler == null) {
+					throw e;
+				}
+				try {
+					this.batchErrorHandler.handle(e, records, this.consumer);
+					if (producer != null) {
+						sendOffsetsToTransaction(producer);
 					}
 				}
+				catch (RuntimeException ee) {
+					this.logger.error("Error handler threw an exception", ee);
+					return ee;
+				}
+				catch (Error er) { //NOSONAR
+					this.logger.error("Error handler threw an error", er);
+					throw er;
+				}
+			}
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
 			}
+			return null;
 		}
 
 		private void invokeRecordListener(final ConsumerRecords<K, V> records) {
+			if (this.transactionTemplate != null) {
+				innvokeRecordListenerInTx(records);
+			}
+			else {
+				doInvokeWithRecords(records);
+			}
+		}
+
+		/**
+		 * Invoke the listener with each record in a separate transaction.
+		 * @param records the records.
+		 */
+		@SuppressWarnings({ "rawtypes" })
+		private void innvokeRecordListenerInTx(final ConsumerRecords<K, V> records) {
 			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
 			while (iterator.hasNext()) {
 				final ConsumerRecord<K, V> record = iterator.next();
 				if (this.logger.isTraceEnabled()) {
 					this.logger.trace("Processing " + record);
 				}
 				try {
-					switch (this.listenerType) {
-						case ACKNOWLEDGING_CONSUMER_AWARE:
-							this.listener.onMessage(record,
-									this.isAnyManualAck
-											? new ConsumerAcknowledgment(record)
-											: null, this.consumer);
-							break;
-						case CONSUMER_AWARE:
-							this.listener.onMessage(record, this.consumer);
-							break;
-						case ACKNOWLEDGING:
-							this.listener.onMessage(record,
-									this.isAnyManualAck
-											? new ConsumerAcknowledgment(record)
-											: null);
-							break;
-						case SIMPLE:
-							this.listener.onMessage(record);
-							break;
-					}
-					if (!this.isAnyManualAck && !this.autoCommit) {
-						this.acks.add(record);
-					}
+					this.transactionTemplate.execute(s -> {
+						Producer producer = null;
+						if (this.kafkaTxManager != null) {
+							producer = ((KafkaResourceHolder) TransactionSynchronizationManager
+									.getResource(this.kafkaTxManager.getProducerFactory())).getProducer();
```

Comment:
Well, either way we don't have `producer` and, therefore, don't `sendOffsetsToTransaction()`.
So, I don't know what "sync'd" means here then...
Or am I missing anything else?

I think we can somehow reuse `ProducerFactoryUtils.getTransactionalResourceHolder()` for any other TX Managers.

But I agree that it might be postponed to separate issue.